README
BFGMiner:
St. Barbara's Faithfully Glorified Mining Initiative Naturally Exceeding Rivals
or Basically a Freaking Good Miner

This is a multi-threaded, multi-blockchain, multi-pool ASIC, FPGA, GPU and CPU
miner with dynamic clocking, monitoring, and fanspeed support for bitcoin.

This code is provided entirely free of charge by the programmer in his spare
time so donations would be greatly appreciated. Please consider donating to the
address below.

Luke-Jr <luke-jr+bfgminer@utopios.org>
1QATWksNFGeUJCWBrN4g6hGM178Lovm7Wh

DOWNLOADS:

http://luke.dashjr.org/programs/bitcoin/files/bfgminer

GIT TREE:

https://github.com/luke-jr/bfgminer

Bug reports:

https://github.com/luke-jr/bfgminer/issues

IRC Channel:

irc://irc.freenode.net/eligius

License: GPLv3.  See COPYING for details.

SEE ALSO README.ASIC, README.FPGA, README.GPU, README.RPC, AND README.scrypt FOR
MORE INFORMATION ON EACH.

---

EXECUTIVE SUMMARY ON USAGE:

Single pool:

bfgminer -o http://pool:port -u username -p password

Multiple pools:

bfgminer -o http://pool1:port -u pool1username -p pool1password -o http://pool2:port -u pool2usernmae -p pool2password

Multiple blockchains:

bfgminer -o http://pool1:port -u pool1username -p pool1password --pool-goal default -o http://pool2:port -u pool2usernmae -p pool2password --pool-goal freicoin

Single pool with a standard http proxy:

bfgminer -o http://pool:port -x http://proxy:port -u username -p password

Single pool with a socks5 proxy:

bfgminer -o http://pool:port -x socks5://proxy:port -u username -p password

The list of proxy types are:
 http:    standard http 1.1 proxy
 socks4:  socks4 proxy
 socks5:  socks5 proxy
 socks4a: socks4a proxy
 socks5h: socks5 proxy using a hostname

Proxy support requires cURL version 7.21.7 or newer.

If you specify the --socks-proxy option to BFGMiner, it will only be applied to
all pools that don't specify their own proxy setting like above


After saving configuration from the menu ([S],[W]) you do not need to give
BFGMiner any arguments, it will load your configuration instead.

Any configuration file may also contain a single
	"include" : "filename"
to recursively include another configuration file.
Writing the configuration will save all settings from all files to the output
configuration file.


---
BUILDING BFGMINER

Everything you probably want, condensed:
	build-essential autoconf automake libtool pkg-config libcurl4-gnutls-dev
	libjansson-dev uthash-dev libncursesw5-dev libudev-dev libusb-1.0-0-dev
	libevent-dev libmicrohttpd-dev libhidapi-dev

Dependencies:
	autoconf             http://www.gnu.org/software/autoconf/
	automake             http://www.gnu.org/software/automake/
	libtool              http://www.gnu.org/software/libtool/
	pkg-config           http://www.freedesktop.org/wiki/Software/pkg-config
	...or pkgconf        https://github.com/pkgconf/pkgconf

	libcurl4-gnutls-dev  http://curl.haxx.se/libcurl/

	libjansson-dev 2.0+  http://www.digip.org/jansson/
	
	uthash-dev 1.9.4+    http://troydhanson.github.io/uthash/

Optional Dependencies:
	Text-User-Interface (TUI): curses dev library; any one of:
	  libncurses5-dev    http://www.gnu.org/software/ncurses/ (Linux and Mac)
	  libncursesw5-dev       ^ same
	  libpdcurses        http://pdcurses.sourceforge.net/ (Linux/Mac/Windows)

	Multiple ASIC/FPGA autodetection: any one of:
	  sysfs              (built-in to most Linux kernels, just mount on /sys)
	  libudev-dev        http://www.freedesktop.org/software/systemd/libudev/
	
	HashBuster Nano & NanoFury USB devices:
	  libhidapi-dev      https://github.com/signal11/hidapi
	
	getwork server for Block Erupter Blades:
	  libmicrohttpd-dev 0.9.5+  http://www.gnu.org/software/libmicrohttpd/
	
	Stratum proxy:
	  libevent 2.0.3+    http://libevent.org/

	KnCMiner SHA2 miners:
	  libi2c-dev         https://i2c.wiki.kernel.org/index.php/I2C_Tools
	
	HashBuster Micro, Klondike, X6500 and ZTEX FPGA boards:
	  libusb-1.0-0-dev   http://www.libusb.org/

	Video card GPU mining (free):
	  llvm 3.3+          http://llvm.org/
	  clang 3.3+         http://clang.llvm.org/
	  libclc             http://libclc.llvm.org/
	  Mesa 9.2.0+        http://www.mesa3d.org/
	  libsensors4-dev    https://github.com/groeck/lm-sensors
	
	ATi/AMD video card GPU mining (non-free):
	  AMD APP SDK        http://developer.amd.com/tools/heterogeneous-computing/amd-accelerated-parallel-processing-app-sdk/

	CPU mining optimized assembly algorithms:
	  yasm 1.0.1+        http://yasm.tortall.net/


BFGMiner driver configuration options:
	--disable-other-drivers Build without drivers by default unless explicitly
	                        enabled
	--enable-broad-udevrules
	                        Include udev rules for ambiguous devices which may
	                        not be miners
	--enable-alchemist      Compile support for AlcheMist (default disabled)
	--disable-avalon        Compile support for Avalon (default enabled)
	--disable-avalonmm      Compile support for Avalon2/3 (default enabled)
	--enable-bfsb           Compile support for BFSB (default disabled)
	--disable-bfx           Compile support for BFx2 (default enabled)
	--disable-bifury        Compile support for Bi*Fury (default enabled)
	--disable-bigpic        Compile support for Big Picture Mining USB (default
	                        enabled)
	--disable-bitforce      Compile support for BitForce (default enabled)
	--disable-bitfury       Compile support for Bitfury (default enabled)
	--enable-bitmain        Compile support for Bitmain Antminer S* series
	                        (default disabled)
	--disable-cointerra     Compile support for CoinTerra (default enabled)
	--enable-cpumining      Compile support for CPU mining (default disabled)
	--disable-drillbit      Compile support for DrillBit (default enabled)
	--disable-dualminer     Compile support for DualMiner (default enabled)
	--disable-gridseed      Compile support for GridSeed (default enabled)
	--disable-hashbuster    Compile support for HashBuster Nano (default
	                        enabled)
	--disable-hashbusterusb Compile support for HashBuster Micro (default
	                        enabled)
	--disable-hashfast      Compile support for HashFast (default enabled)
	--disable-icarus        Compile support for Icarus (default enabled)
	--enable-jingtian       Compile support for JingTian (default disabled)
	--disable-klondike      Compile support for Klondike (default enabled)
	--enable-knc            Compile support for KnC (default disabled)
	--enable-kncasic        Compile support for KnC gen 2 (default disabled)
	--disable-littlefury    Compile support for LittleFury (default enabled)
	--enable-metabank       Compile support for Metabank (default disabled)
	--enable-minergate      Compile support for Spondoolies minergate interface
	                        (default disabled)
	--enable-minion         Compile support for Minion (default disabled)
	--disable-modminer      Compile support for ModMiner (default enabled)
	--disable-nanofury      Compile support for NanoFury (default enabled)
	--enable-opencl         Compile support for OpenCL (default disabled)
	--disable-adl           Build without ADL monitoring (default enabled)
	--disable-rockminer     Compile support for RockMiner (default enabled)
	--enable-titan          Compile support for KnC Titan (default disabled)
	--disable-twinfury      Compile support for Twinfury (default enabled)
	--disable-x6500         Compile support for X6500 (default enabled)
	--disable-zeusminer     Compile support for ZeusMiner (default enabled)
	--disable-ztex          Compile support for ZTEX (default enabled)

BFGMiner algorithm configuration option:
	--enable-keccak         Compile support for Keccak (default disabled)
	--disable-sha256d       Compile support for SHA256d (default enabled)
	--enable-scrypt         Compile support for scrypt (default disabled)

BFGMiner dependency configuration options:
	--without-curses        Compile support for curses TUI (default enabled)
	--without-libevent      Compile support for libevent stratum server (default
	                        enabled)
	--without-libmicrohttpd Compile support for libmicrohttpd getwork server
	                        (default enabled)
	--without-libudev       Autodetect FPGAs using libudev (default enabled)
	--without-libusb        Compile using libusb (default enabled)
	--without-sensors       Build with libsensors monitoring (default enabled)
	--with-system-libblkmaker
	                        Use system libblkmaker rather than bundled one
	                        (default disabled)
	--with-udevrulesdir=DIR Install udev rules into this directory
	--with-udevrules-group=groupname
	                        Configure mining devices to be owned by a specific
	                        group (default `video')
	--without-uio           Compile support for PCI devices via Linux UIO
	                        interface (default enabled)
	--without-vfio          Compile support for PCI devices via Linux VFIO
	                        interface (default enabled)

Basic *nix build instructions:

./autogen.sh    # only needed if building from git repo
./configure  # list configuration options here
make

No installation is necessary. You may run BFGMiner from the build directory
directly.

On Mac OS X, you can use Homebrew to install the dependency libraries. When you
are ready to build BFGMiner, you may need to point the configure script at one
or more pkg-config paths. For example:
	./configure PKG_CONFIG_PATH=/usr/local/opt/curl/lib/pkgconfig:/usr/local/opt/jansson/lib/pkgconfig

Native WIN32 build instructions: see windows-build.txt

If you build BFGMiner from source, it is recommended that you run it from the
build directory. On *nix, you will usually need to prepend your command with a
path like this (if you are in the bfgminer directory already): ./bfgminer
To install system wide run 'sudo make install' or 'make install' as root. You
can then run from any terminal.

---

Usage instructions:  Run "bfgminer --help" to see options:

Usage: bfgminer [-DdElmpPQqUsTouOchnV]

Options for both config file and command line:
--api-allow         Allow API access (if enabled) only to the given list of [W:]IP[/Prefix] address[/subnets]
                    This overrides --api-network and you must specify 127.0.0.1 if it is required
                    W: in front of the IP address gives that address privileged access to all api commands
--api-description   Description placed in the API status header (default: BFGMiner version)
--api-groups        API one letter groups G:cmd:cmd[,P:cmd:*...]
                    See README.RPC for usage
--api-listen        Listen for API requests (default: disabled)
                    By default any command that does not just display data returns access denied
                    See --api-allow to overcome this
--api-mcast         Enable API Multicast listener, default: disabled
--api-mcast-addr <arg> API Multicast listen address (default: "224.0.0.75")
--api-mcast-code <arg> Code expected in the API Multicast message, don't use '-' (default: "FTW")
--api-mcast-des <arg>  Description appended to the API Multicast reply, default: ''
--api-mcast-port <arg> API Multicast listen port (default: 4028)
--api-network       Allow API (if enabled) to listen on/for any address (default: only 127.0.0.1)
--api-port          Port number of miner API (default: 4028)
--balance           Change multipool strategy from failover to even share balance
--benchmark         Run BFGMiner in benchmark mode - produces no shares
--benchmark-intense Run BFGMiner in intensive benchmark mode - produces no shares
--chroot-dir <arg>  Chroot to a directory right after startup
--cmd-idle <arg>    Execute a command when a device is allowed to be idle (rest or wait)
--cmd-sick <arg>    Execute a command when a device is declared sick
--cmd-dead <arg>    Execute a command when a device is declared dead
--coinbase-check-addr <arg> A list of address to check against in coinbase payout list received from the previous-defined pool, separated by ','
--coinbase-check-total <arg> The least total payout amount expected in coinbase received from the previous-defined pool
--coinbase-check-percent <arg> The least benefit percentage expected for the sum of addr(s) listed in --cbaddr argument for previous-defined pool
--coinbase-sig <arg> Set coinbase signature when possible
--compact           Use compact display without per device statistics
--debug|-D          Enable debug output
--debuglog          Enable debug logging
--device-protocol-dump Verbose dump of device protocol-level activities
--device|-d <arg>   Enable only devices matching pattern (default: all)
--disable-rejecting Automatically disable pools that continually reject shares
--http-port <arg>   Port number to listen on for HTTP getwork miners (-1 means disabled) (default: -1)
--expiry <arg>      Upper bound on how many seconds after getting work we consider a share from it stale (w/o longpoll active) (default: 120)
--expiry-lp <arg>   Upper bound on how many seconds after getting work we consider a share from it stale (with longpoll active) (default: 3600)
--failover-only     Don't leak work to backup pools when primary pool is lagging
--failover-switch-delay <arg> Delay in seconds before switching back to a failed pool (default: 300)
--generate-to <arg> Set an address to generate to for solo mining
--force-dev-init    Always initialize devices when possible (such as bitstream uploads to some FPGAs)
--kernel-path <arg> Specify a path to where bitstream and kernel files are
--load-balance      Change multipool strategy from failover to quota based balance
--log|-l <arg>      Interval in seconds between log output (default: 20)
--log-file|-L <arg> Append log file for output messages
--log-microseconds  Include microseconds in log output
--monitor|-m <arg>  Use custom pipe cmd for output messages
--net-delay         Impose small delays in networking to avoid overloading slow routers
--no-gbt            Disable getblocktemplate support
--no-getwork        Disable getwork support
--no-hotplug        Disable hotplug detection
--no-local-bitcoin  Disable adding pools for local bitcoin RPC servers
--no-longpoll       Disable X-Long-Polling support
--no-pool-redirect  Ignore pool requests to redirect to another server
--no-restart        Do not attempt to restart devices that hang
--no-stratum        Disable Stratum detection
--no-submit-stale   Don't submit shares if they are detected as stale
--no-unicode        Don't use Unicode characters in TUI
--noncelog <arg>    Create log of all nonces found
--pass|-p <arg>     Password for bitcoin JSON-RPC server
--per-device-stats  Force verbose mode and output per-device statistics
--pool-goal <arg>   Named goal for the previous-defined pool
--pool-priority <arg> Priority for just the previous-defined pool
--pool-proxy|-x     Proxy URI to use for connecting to just the previous-defined pool
--protocol-dump|-P  Verbose dump of protocol-level activities
--queue|-Q <arg>    Minimum number of work items to have queued (0 - 10) (default: 1)
--quiet|-q          Disable logging output, display status and errors
--quit-summary <arg> Summary printed when you quit: none/devs/procs/detailed
--quota|-U <arg>    quota;URL combination for server with load-balance strategy quotas
--real-quiet        Disable all output
--request-diff <arg> Request a specific difficulty from pools (default: 1.0)
--retries <arg>     Number of times to retry failed submissions before giving up (-1 means never) (default: -1)
--rotate <arg>      Change multipool strategy from failover to regularly rotate at N minutes (default: 0)
--round-robin       Change multipool strategy from failover to round robin on failure
--scan|-S <arg>     Configure how to scan for mining devices
--scan-time <arg>   Upper bound on time spent scanning current work, in seconds (default: 60)
--sched-start <arg> Set a time of day in HH:MM to start mining (a once off without a stop time)
--sched-stop <arg>  Set a time of day in HH:MM to stop mining (will quit without a start time)
--scrypt            Use the scrypt algorithm for mining (non-bitcoin)
--set-device|--set <arg> Set default parameters on devices; eg, NFY:osc6_bits=50, bfl:voltage=<value>, compac:clock=<value>
--setuid <arg>      Username of an unprivileged user to run as
--sharelog <arg>    Append share log to file
--shares <arg>      Quit after mining 2^32 * N hashes worth of shares (default: unlimited)
--show-processors   Show per processor statistics in summary
--skip-security-checks <arg> Skip security checks sometimes to save bandwidth; only check 1/<arg>th of the time (default: never skip)
--socks-proxy <arg> Set socks proxy (host:port) for all pools without a proxy specified
--stratum-port <arg> Port number to listen on for stratum miners (-1 means disabled) (default: -1)
--submit-threads    Minimum number of concurrent share submissions (default: 64)
--syslog            Use system log for output messages (default: standard error)
--temp-hysteresis <arg> Set how much the temperature can fluctuate outside limits when automanaging speeds (default: 3)
--text-only|-T      Disable ncurses formatted screen output
--unicode           Use Unicode characters in TUI
--url|-o <arg>      URL for bitcoin JSON-RPC server
--user|-u <arg>     Username for bitcoin JSON-RPC server
--verbose           Log verbose output to stderr as well as status output
--weighed-stats     Display statistics weighed to difficulty 1
--userpass|-O <arg> Username:Password pair for bitcoin JSON-RPC server
--worktime                     Display extra work time debug information
Options for command line only:
--config|-c <arg>   Load a JSON-format configuration file
See example.conf for an example configuration.
--no-default-config Inhibit loading default config file
--default-config    Always load the default config file
--help|-h           Print this message
--version|-V        Display version and exit


GPU only options:

--gpu-map <arg>     Map OpenCL to ADL device order manually, paired CSV (e.g. 1:0,2:1 maps OpenCL 1 to ADL 0, 2 to 1)
--gpu-platform <arg> Select OpenCL platform ID to use for GPU mining
--gpu-reorder       Attempt to reorder GPU devices according to PCI Bus ID
--no-adl            Disable the ATI display library used for monitoring and setting GPU parameters

GPU mining is disabled by default for SHA256d if you have any dedicated mining
devices, but can be enabled explicitly specifying the -S opencl:auto option.

See README.GPU for more information regarding GPU mining.
See README.scrypt for more information regarding (non-bitcoin) scrypt mining.


To use ASICs or FPGAs, you will need to be sure the user BFGMiner is running as
has appropriate permissions. This varies by operating system.
On Linux, with BFGMiner's udev rules: sudo usermod <username> -a -G video
Note that on GNU/Linux systems, you will usually need to login again before
group changes take effect.

By default, BFGMiner will scan for autodetected devices. If you want to prevent
BFGMiner from doing this, you can use "-S noauto". If you want to probe all
serial ports, you can use "-S all"; note that this may write data to non-mining
devices which may then behave in unexpected ways!

On Linux, <arg> is usually of the format /dev/ttyUSBn
On Mac OS X, <arg> is usually of the format /dev/cu.usb*
On Windows, <arg> is usually of the format \\.\COMn
(where n = the correct device number for the device)

The official supplied binaries are compiled with support for all ASICs/FPGAs.
To force the code to only attempt detection with a specific driver,
prepend the argument with the driver name followed by an "at" symbol.
For example, "icarus@/dev/ttyUSB0" or "bitforce@\\.\COM5"
or using the short name: "ica@/dev/ttyUSB0" or "bfl@\\.\COM5"

Some FPGAs do not have non-volatile storage for their bitstreams and must be
programmed every power cycle, including first use. To use these devices, you
must download the proper bitstream from the vendor's website and copy it to the
"bitstreams" directory into your BFGMiner application directory.

See README.ASIC and README.FPGA for more information regarding these.

See README.CPU for information regarding CPU mining.

---

WHILE RUNNING:

The following options are available while running with a single keypress:

[M]anage devices [P]ool management [S]ettings [D]isplay options  [H]elp [Q]uit

M gives you something like:

Select processor to manage using up/down arrow keys
 BFL 0a: 78.0C |  3.64/ 3.70/ 2.91Gh/s | A:46 R:0+0(none) HW:  2/none
  BitFORCE SHA256 SC from Butterfly Labs
Serial: FTWN6T67

[D]isable
Or press Enter when done


P gives you:

Current pool management strategy: Failover
[F]ailover only disabled
[A]dd pool [R]emove pool [D]isable pool [E]nable pool
[C]hange management strategy [S]witch pool [I]nformation


S gives you:

[L]ongpoll: On
[Q]ueue: 1
[S]cantime: 60
[E]xpiry: 120
[R]etries: -1
[W]rite config file
[B]FGMiner restart


D gives you:

[N]ormal [C]lear [S]ilent mode (disable all output)
[D]ebug:off
[P]er-device:off
[Q]uiet:off
[V]erbose:off
[R]PC debug:off
[W]orkTime details:off
co[M]pact: off
[L]og interval:5


Q quits the application.


The running log shows output similar to that below:

 [2013-02-13 00:26:30] Accepted 1758e8df BFL 0  pool 0 Diff 10/1
 [2013-02-13 00:26:32] Accepted 1d9a2199 MMQ 0a pool 0 Diff 8/1
 [2013-02-13 00:26:33] Accepted b1304924 ZTX 0  pool 0 Diff 1/1
 [2013-02-13 00:26:33] Accepted c3ad22f4 XBS 0b pool 0 Diff 1/1

The 8 byte hex value are the 2nd set of 32 bits from the share submitted to the
pool. The 2 diff values are the actual difficulty target that share reached
followed by the difficulty target the pool is currently asking for.

---
Also many issues and FAQs are covered in the forum threads
dedicated to this program,
	https://bitcointalk.org/?topic=78192
	https://bitcointalk.org/?topic=168174

If you are mining on a single pool, the pool display shows:
 Pool 0: ...s.m.eligius.st   Diff:16  +Strtm  LU:[03:26:16]  User:1QATWksNFGeUJCWBrN4g6hGM178Lovm7Wh

This tells you which pool you're using, as well as its current share difficulty,
protocol, and last explicit work update. If BFGMiner has a working block
notification source, the protocol will be prefixed by a plus sign. If not, a
minus sign.

If you are mining on multiple pools at once, the pool display instead shows:
 Pools: 2 (0,1)              Diff:4-16  +  LU:[03:25:30]

You get the total number of working pools, the pool numbers for each of those,
the range of current share difficulties, whether block notification is working
(plus/minus), and the oldest explicit work update currently being used for new
work.

The block display shows:
Block #217364: ...1b89f8d3  Diff:7.67M (54.93T)  Started: [17:17:22]  I:12.99mBTC/hr

This shows a short stretch of the next block's height, the current block,
difficulty (including the network hashrate that difficulty represents), when the
search for the new block started, and finally expected Income, calculated by
actual shares submitted in 100% PPS value (assumes Bitcoin, does not account for
altcoin conversions!).

The BFGMiner status line shows:
 ST:1  F:0  NB:1  AS:0  BW:[ 75/241 B/s]  E:2.42  BS:2.71k

ST is STaged work items (ready to use).
F  is network Failure occasions (server down or slow to provide work)
NB is New Blocks detected on the network
AS is Active Submissions (shares in the process of submitting)
BW is BandWidth usage on the network (received/sent)
E  is Efficiency defined as number of shares accepted (multiplied by their
          difficulty) per 2 KB of bandwidth
BS is the all time Best Share difficulty you've found

The totals line shows the following:
 6/32   75.0C | 171.3/170.8/171.2Gh/s | A:729 R:8+0(.01%) HW:0/.81%

Each column is as follows:
  The number of devices and processors currently mining
  Hottest temperature reported by any processor
  20 second exponentially decaying average hash rate (configurable with --log
      option)
  An all time average hash rate
  An all time average hash rate based on actual nonces found, adjusted for pool
      reject and stale rate
  The number of Accepted shares
  The number of Rejected shares and stale shares discarded (never submitted),
      and the percentage these are of total found.
  The number of HardWare errors, and percentage invalid of nonces returned

Each device shows:
 BFL 2: 74.0C | 51.97/58.90/57.17Gh/s | A:847 R:15+0(.54%) HW:496/.91%

Columns are the same as in the totals line.


---
MULTIPOOL

FAILOVER STRATEGIES WITH MULTIPOOL:
A number of different strategies for dealing with multipool setups are
available. Each has their advantages and disadvantages so multiple strategies
are available by user choice, as per the following list:

FAILOVER:
The default strategy is failover. This means that if you input a number of
pools, it will try to use them as a priority list, moving away from the 1st
to the 2nd, 2nd to 3rd and so on. If any of the earlier pools recover, it will
move back to the higher priority ones.

ROUND ROBIN:
This strategy only moves from one pool to the next when the current one falls
idle and makes no attempt to move otherwise.

ROTATE:
This strategy moves at user-defined intervals from one active pool to the next,
skipping pools that are idle.

LOAD BALANCE:
This strategy sends work to all the pools on a quota basis. By default, all
pools are allocated equal quotas unless specified with --quota. This
apportioning of work is based on work handed out, not shares returned so is
independent of difficulty targets or rejected shares. While a pool is disabled
or dead, its quota is dropped until it is re-enabled. Quotas are forward
looking, so if the quota is changed on the fly, it only affects future work.
If all pools are set to zero quota or all pools with quota are dead, it will
fall back to a failover mode. See quota below for more information.

The failover-only flag has special meaning in combination with load-balance
mode and it will distribute quota back to priority pool 0 from any pools that
are unable to provide work for any reason so as to maintain quota ratios
between the rest of the pools.

BALANCE:
This strategy monitors the amount of difficulty 1 shares solved for each pool
and uses it as a basis for trying to doing the same amount of work for each
pool.


---
SOLO MINING

BFGMiner supports solo mining with any GBT-compatible bitcoin node (such as
bitcoind). To use this mode, you need to specify the URL of your bitcoind node
using the usual pool options (--url, --userpass, etc), and the --generate-to
option to specify the Bitcoin address you wish to receive the block rewards
mined. When you run Bitcoin Core on the same computer as your miner, the pool
itself will be automatically configured for you (on the default goal). Please be
aware that solo mining via GBT is at this time only supported for Bitcoin.

IMPORTANT: If you are solo mining with more than one instance of BFGMiner (or
any other software) per payout address, you must also specify data using the
--coinbase-sig option to ensure each miner is working on unique work. Note

NEWS
BFGMiner Version 5.5.0 - January 3, 2018

- Update official Win32/64 build compiler and libraries:
- - Win64 compiler (GCC) from 5.3.0 to 7.2.0 (Win32 remains at 4.7.4)
- - c-ares from 1.10.0-r1 to 1.11.0
- - libjansson from 2.7 to 2.10
- - libcurl from 7.47.1 to 7.57.0
- - libevent from 2.1.5-r4 to 2.1.8
- - libusb from 1.0.20 to 1.0.21
- - libmicrohttpd from 0.9.48 to 0.9.58
- - zlib from 1.2.8-r1 to 1.2.11-r1
- Update copyrights
- AUTHORS: Update
- README.ASIC: Update Alcheminer
- alchemist: Reset using DTR as well
- alchemist: Experimental USB->UART Dongle Support on RPi
- make-release: Add crypt32.dll to OS-provided library list
- README: Expand on --set-device documentation
- Update libblkmaker to 0.6.0
- openwrt: Update for LEDE 17.01
- ccan: Update to latest version (which fixes musl libc support)
- Bugfix: probe_device_thread: Fix threadname buffer size
- Bugfix: appdata_file_call: Fix buffer size and loop
- Bugfix: SSM: Clear desired_default_share_pdiff when initialising
desired_share_pdiff before authentication
- cpu: Avoid relocations for x86_64 assembly that interfere with building
BFGMiner as a PIE
- Bugfix: SSM: We already hold the pool lock when we need to call stale_work,
so avoid trying to lock it a second time inside the function
- Bugfix: Hold the control lock when updating the generate-to script
- Bugfix: Hold the pool lock when updating lp_id
- Bugfix: icarus: Explicitly specify type wide enough for bitshift in
icarus_proc_for_nonce
- configure: BFG_FIND_INCLUDE_PATH: Handle more unusual syntax correctly
- Bugfix: SSM: Handle network problems safely, and use thread-safe libevent


BFGMiner Version 5.4.2 - March 26, 2016

- Update official Win32/64 build compiler and libraries:
- - Win64 compiler (GCC) from 4.7.4 to 5.3.0 (Win32 remains at 4.7.4)
- - libcurl from 7.39.0 to 7.47.1
- - libevent from 2.0.21 to 2.1.5-r4
- - libusb from 1.0.18 (Win32) and 1.0.19 (Win64) to 1.0.20
- - libmicrohttpd from 0.9.38_pre33603 to 0.9.48
- Update libblkmaker to 0.5.3
- Bugfix: Sanitise lock situation for work2d
- Bugfix: SSM: Never issue the same work2d twice in a row
- Bugfix: SSM: Never send mining.notify for stale work2d
- SSM: Refactor work2d check and job pruning from _stratumsrv_update_notify
into stratumsrv_update_notify_str
- Updates for OpenWrt 15.05
- Bugfix: Correct work2d handling for GBT servers


BFGMiner Version 5.4.1 - November 21, 2015

- bitmain: Use more portable PRIu64 rather than %llu (which fails on Windows)
- icarus: Use more portable PRIu64 rather than %llu (which fails on Windows)
- Bugfix: openwrt: Fix bitmain option
- Bugfix: antminer: Avoid strstr(NULL, ...) when device has no [available]
product string
- Update libblkmaker to 0.5.2
- openwrt: Add option to enable bitmain driver
- openwrt: Update libusb device list


BFGMiner Version 5.4.0 - October 23, 2015

- AUTHORS: Move jstefanop to current maintainers
- Disable alchemist by default, add Travis build, and document configure option
in README
- alchemist: New scrypt ASIC driver
- hashfast: Support setting clock speed from TUI Manage
- avalon: Support setting clock speed from TUI Manage
- antminer: Support setting clock speed from TUI Manage (hidden for non-BM1382/4
devices)
- jingtian: Allow changing clock speed from TUI
- Debuglog for retrodiff being disabled
- Disable retrodiff when it was not explicitly enabled, and the pool doesn't
seem to like it
- Bugfix: Only adjust work_difficulty on retrodiff submissions, so we can still
detect them on rejection
- (Re-)enable retrodiff by default for stratum pools, since some servers
implement mining.set_difficulty wrong in this way
- New pool option "retrodiff"
- If a share passes difficulty check for the updated pool target,   record that
target for correct accounting of diff_stale.
- DevAPI: Remove unused temporary variable from driver iteration
- README.ASIC: Expand Antminer S5 section to S1-S5 for completeness, since in
theory they should work
- bitmain: Make reg_data optional for S4 and S5, calculating it from clock
- Document Bitmain Antminer S5 support
- bitmain: Remove poll_prio_threshold and just ensure we poll at a regular
interval to avoid nonce buffer overruns
- bitmain: Remove dead per-device constants
- bitmain: Only have one set of actual maximums
- bitmain: Migrate BITMAIN_MAX_NONCE_NUM to runtime packet_max_nonce
- bitmain: Migrate BITMAIN_MAX_WORK_QUEUE_NUM/bitmain_work_poll_prio to runtime
poll_prio_threshold
- bitmain: Migrate BITMAIN_MAX_WORK_NUM to runtime packet_max_work
- bitmain: Dummy model parameter
- bitmain: Remove baud/flush, as the fake VCOM kernel driver ignores (and
complains about) both
- bitmain: Poll after queuing work, once enough is queued
- bitmain: Sleep a little between polling
- bitmain: Flush queues for work restarts
- bitmain: Queue multiple works at a time
- bitmain: Remove more dead code
- bitmain: Prune old work eventually
- Bugfix: bitmain: Properly count work-search failures as HW errors
- bitmain: Refactor to work with minerloop_queue (leaks work memory)
- bitmain: Avoid busy-looping on network devices
- bitmain: Support for talking to a device over the network
- bitmain: Use the hottest reading for device temperature
- Bugfix: bitmain: Actually use baud setting
- bitmain: Remove more dead code
- bitmain: Handle stale results normally
- bitmain: Simplify num2bit
- bitmain: Initial set-device based configuration
- bitmain: Use lowl-vcom abstractions
- Update copyrights
- Bugfix: bitmain: Type-safe printf-format usage
- Travis: Test bitmain-only build
- Build bitmain driver
- bitmain: Hardcode configuration for now
- bitmain: Rework nonce_diff handling
- bitmain: Update headers
- bitmain: Deal with unused variable warnings sanely
- bitmain: Explicitly de-const device_path for freeing
- Bugfix: bitmain: Correct type of cgpu->temp reference
- Bugfix: bitmain: Use uint8_t consistently for binary data
- Bugfix: bitmain: Clear opt_bitmain_dev string properly
- bitmain: Remove non-existent bitmain_detect parameter

sha2.h
/*
 * FIPS 180-2 SHA-224/256/384/512 implementation
 * Last update: 02/02/2007
 * Issue date:  04/30/2005
 *
 * Copyright 2013 Con Kolivas
 * Copyright 2005, 2007 Olivier Gay
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the project nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include "config.h"

#include <stdint.h>

#include "miner.h"

#ifndef SHA2_H
#define SHA2_H

#define SHA256_DIGEST_SIZE ( 256 / 8)
#define SHA256_BLOCK_SIZE  ( 512 / 8)

#define SHFR(x, n)    (x >> n)
#define ROTR(x, n)   ((x >> n) | (x << ((sizeof(x) << 3) - n)))
#define CH(x, y, z)  ((x & y) ^ (~x & z))
#define MAJ(x, y, z) ((x & y) ^ (x & z) ^ (y & z))

#define SHA256_F1(x) (ROTR(x,  2) ^ ROTR(x, 13) ^ ROTR(x, 22))
#define SHA256_F2(x) (ROTR(x,  6) ^ ROTR(x, 11) ^ ROTR(x, 25))
#define SHA256_F3(x) (ROTR(x,  7) ^ ROTR(x, 18) ^ SHFR(x,  3))
#define SHA256_F4(x) (ROTR(x, 17) ^ ROTR(x, 19) ^ SHFR(x, 10))

typedef struct {
    unsigned int tot_len;
    unsigned int len;
    unsigned char block[2 * SHA256_BLOCK_SIZE];
    uint32_t h[8];
} sha256_ctx;

extern uint32_t sha256_k[64];

void sha256_init(sha256_ctx * ctx);
void sha256_update(sha256_ctx *ctx, const unsigned char *message,
                   unsigned int len);
void sha256_final(sha256_ctx *ctx, unsigned char *digest);
void sha256(const unsigned char *message, unsigned int len,
            unsigned char *digest);

#endif /* !SHA2_H */
